framework/src/org/checkerframework/framework/util/AtmLubVisitor.java

```diff
@@ -0,0 +1,365 @@
+package org.checkerframework.framework.util;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.type.TypeVariable;
+import org.checkerframework.framework.type.AnnotatedTypeFactory;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedArrayType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedDeclaredType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedIntersectionType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedNullType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedPrimitiveType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedTypeVariable;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedUnionType;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedWildcardType;
+import org.checkerframework.framework.type.QualifierHierarchy;
+import org.checkerframework.framework.type.visitor.AbstractAtmComboVisitor;
+import org.checkerframework.framework.type.visitor.VisitHistory;
+import org.checkerframework.javacutil.ErrorReporter;
+import org.checkerframework.javacutil.InternalUtils;
+import org.checkerframework.javacutil.TypesUtils;
+
+/**
+ * Helper class to compute the least upper bound of two AnnotatedTypeMirrors.
+ *
+ * This class should only be used by
+ * {@link AnnotatedTypes#leastUpperBound(AnnotatedTypeFactory, AnnotatedTypeMirror, AnnotatedTypeMirror)}.
+ */
+class AtmLubVisitor extends AbstractAtmComboVisitor<Void, AnnotatedTypeMirror> {
+
+    private final AnnotatedTypeFactory atypeFactory;
+    private final QualifierHierarchy qualifierHierarchy;
+    private VisitHistory visitHistory = new VisitHistory();
+
+    AtmLubVisitor(AnnotatedTypeFactory atypeFactory) {
+        this.atypeFactory = atypeFactory;
+        this.qualifierHierarchy = atypeFactory.getQualifierHierarchy();
+    }
+
+    /**
+     * Returns an ATM that is the least upper bound of type1 and type2 and whose Java type is
+     * lubJavaType.  lubJavaType must be a super type or convertible to the Java types of type1 and
+     * type2.
+     **/
+    AnnotatedTypeMirror lub(
+            AnnotatedTypeMirror type1, AnnotatedTypeMirror type2, TypeMirror lubJavaType) {
+        visitHistory = new VisitHistory();
```

Comment:
Doesn't this cause visitHistory to be set once from the field initializer and then again immediately here, without ever using the first value?
How about adding a `clear` method to VisitHistory, making the field final, and calling clear before returning?